src/Stack/Setup.hs

```diff
@@ -669,6 +674,91 @@ ensureCompiler sopts = do
       , cpGlobalDB' = globaldb
       }
 
+buildGhcFromSource :: forall env.
+   ( HasTerm env
+   , HasProcessContext env
+   , HasConfig env
+   ) => Memoized SetupInfo -> [Tool] -> CompilerRepository -> Text -> Text
+   -> RIO env (Either (ActualCompiler, Arch, Path Abs Dir) Tool, CompilerBuild)
+buildGhcFromSource getSetupInfo' installed (CompilerRepository url) commitId flavour = do
+   config <- view configL
+   let compilerTool = ToolGhcGit commitId flavour
+
+   -- detect when the correct GHC is already installed
+   if compilerTool `elem` installed
+     then return (Right compilerTool,CompilerBuildStandard)
+     else do
+       let repo = Repo
+            { repoCommit = commitId
+            , repoUrl    = url
+            , repoType   = RepoGit
+            , repoSubdir = mempty
+            }
+
+       -- clone the repository and execute the given commands
+       Pantry.withRepo repo $ do
+         threads <- view $ configL.to configJobs
+         let
+           hadrianArgs = fmap T.unpack
+               [ hadrianCmd
+               , "-c"                    -- run ./boot and ./configure
+               , "-j" <> tshow threads   -- parallel build
+               , "--flavour=" <> flavour -- selected flavour
+               , "binary-dist"
+               ]
+           hadrianCmd
+             | osIsWindows = "./hadrian/build.stack.bat"
+             | otherwise   = "./hadrian/build.stack.sh"
+
+         logSticky $ "Building GHC from source with `"
+            <> RIO.display flavour
+            <> "` flavour. It can take a long time (more than one hour)..."
+
+         -- RIO.Process doesn't wrap process' "shell".
+         -- Instead we use "proc" with the "sh" command
+         proc "sh" hadrianArgs runProcess_
+           `catch` \(e :: ExitCodeException) -> do
```

Comment:
I think all of this is unnecessary now, since the stdout and stderr will be displayed to the user while the process is running.